ap/tasks/histograms.py

```diff
@@ -187,62 +188,89 @@ class MergeHistograms(
+        "missing ones; default: False",
+    )
+    remove_previous = luigi.BoolParameter(
+        default=False,
+        significant=False,
+        description="when True, remove particlar input histograms after merging; default: False",
+    )
+
     sandbox = dev_sandbox("bash::$AP_BASE/sandboxes/venv_columnar.sh")
 
     shifts = set(CreateHistograms.shifts)
 
-    # in each step, merge 10 into 1
-    merge_factor = 10
-
     # default upstream dependency task classes
     dep_CreateHistograms = CreateHistograms
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    def create_branch_map(self):
+        # create a dummy branch map so that this task could as a job
+        return {0: None}
 
-        # tell ForestMerge to not cache the internal merging structure by default,
-        # (this is enabled in merge_workflow_requires)
-        self._cache_forest = False
+    def workflow_requires(self, only_super: bool = False):
+        reqs = super().workflow_requires()
+        if only_super:
+            return reqs
 
-    def create_branch_map(self):
-        # DatasetTask implements a custom branch map, but we want to use the one in ForestMerge
-        return law.tasks.ForestMerge.create_branch_map(self)
+        reqs["hists"] = self.as_branch().requires()
 
-    def merge_workflow_requires(self):
-        req = self.dep_CreateHistograms.req(self, _exclude={"branches"})
+        return reqs
 
-        # if the merging stats exist, allow the forest to be cached
-        self._cache_forest = req.merging_stats_exist
+    def requires(self):
+        # optional dynamic behavior: determine not yet created variables and require only those
+        prefer_cli = {"variables"}
+        variables = self.variables
+        if self.only_missing:
+            prefer_cli.clear()
+            missing = self.output().count(existing=False, keys=True)[1]
+            variables = tuple(sorted(missing, key=variables.index))
+            if not variables:
+                return []
+
+        return self.dep_CreateHistograms.req(
+            self,
+            branch=-1,
+            variables=tuple(variables),
+            _exclude={"branches"},
+            _prefer_cli=prefer_cli,
+        )
 
-        return req
+    def output(self):
+        return law.SiblingFileCollection({
+            variable_name: self.target(f"hist__{variable_name}.pickle")
+            for variable_name in self.variables
+        })
 
-    def merge_requires(self, start_leaf, end_leaf):
-        return [
-            self.dep_CreateHistograms.req(self, branch=i)
-            for i in range(start_leaf, end_leaf)
+    def run(self):
+        # preare inputs and outputs
+        inputs = self.input()["collection"]
+        outputs = self.output()
+
+        # load input histograms
+        hists = [
+            inp.load(formatter="pickle")
+            for inp in self.iter_progress(inputs.targets.values(), len(inputs), reach=(0, 50))
```

Comment:
what is the 'reach' parameter doing?

Reply:
There is a progress bar in the central scheduler gui that is increased from reach[0] to reach[1] (defaulting to 0 and 100) during the iteration. Since there are two iterations (input + output), each progress iteration is just filling half of the progress.